WorldCat Discovery Service.js

```diff
@@ -0,0 +1,287 @@
+
+	//single result
+	// generate item and return type
+	var co = getFirstContextObj(doc);
+	if(!co) return false;
+
+	return generateItem(doc, co).itemType;
+}
+
+
+/**
+ * Gets Zotero item from a WorldCat icon src
+ */
+function getZoteroType(iconSrc) {
+	// only specify types not specified in COinS
+	if (iconSrc.indexOf("icon-rec") != -1) {
+		return "audioRecording";
+	}
+	if (iconSrc.indexOf("icon-com") != -1) {
+		return "computerProgram";
+	}
+	if (iconSrc.indexOf("icon-map") != -1) {
+		return "map";
+	}
+	return false;
+}
+
+/**
+ * Generates a Zotero item from a single item WorldCat page,
+ * or the first item on a multiple item page
+ */
+function generateItem(doc, co) {
+	var item = new Zotero.Item();
+	ZU.parseContextObject(co, item);
+	// item types not covered by COinS will still need to be covered. Leaving the corresponding code from open Worldcat
+	// here as a reminder.
+
+/*
+	var type = ZU.xpathText(doc,
+		'//img[@class="icn"][contains(@src, "icon-")][1]/@src');
+	if (type) {
+		type = getZoteroType(type);
+		if (type) item.itemType = type;
+	} */
+	
+	return item;
+}
+
+function getSearchResults(doc) {
+	var results = ZU.xpath(doc, '//ol[@class="results"]/li[contains(@id, "record")]')
+	return results;
+}
+
+function getTitleNode(searchResult) {
+	return ZU.xpath(searchResult, './div[@class="title"]/a')[0];
+}
+
+function getFirstContextObj(doc) {
+	return ZU.xpathText(doc, '//span[@class="Z3988"][1]/@title');
+}
+
+
+
+/**
+ * Given an item URL, extract OCLC ID
+ */
+function extractOCLCID(url) {
+	var id = url.match(/\/oclc\/([^?]+)/);
+	if(!id) return false;
+	return id[1];
+}
+
+
+/**
+ * Given an item URL, extract database ID
+ */
+function extractDatabaseID(url) {
+	var id = url.match(/databaseList=[^&#]+/);
+	if(!id) return false;
+	return id[0];
+}
+
+function composeURL(oclcID, databaseID){
+	var risURL =  "/share/citation.ris?oclcNumber=" + oclcID + "&databaseIds=" +databaseID;
+	return risURL;
+	
+}
+/**
+ * RIS Scraper Function
+ *
+ */
+
+function scrape(risURL) {
+	
+	ZU.doGet(risURL, function (text) {
+		//Z.debug(text);
+		//conference proceedings exported as CONF, but fields match BOOK better
+		text = text.replace(/TY\s+-\s+CONF\s+[\s\S]+\n\s*ER\s+-/g, function(m) {
+			return m.replace(/^TY\s+-\s+CONF\s*$/mg, 'TY  - BOOK')
+				//authors are actually editors
+				.replace(/^A1\s+-\s+/mg, 'A3  - ');
+		})
+		
+		Zotero.debug("Importing corrected RIS: \n" + text);
+		
+		var translator = Zotero.loadTranslator("import");
+		translator.setTranslator("32d59d2d-b65a-4da4-b0a3-bdd3cfb979e7");
+		translator.setString(text);
+		translator.setHandler("itemDone", function (obj, item) {
+			item.extra = undefined;
+			item.archive = undefined;
+			//clean up ISBNs
+			if (item.ISBN){
+				var ISBN = item.ISBN.split(/\s/);
+				var ISBNarray = [];
+				for (var i=0; i<ISBN.length; i++){
+					if (ZU.cleanISBN(ISBN[i])){
+						ISBNarray.push(ZU.cleanISBN(ISBN[i]));
+					} 
+				}
+				item.ISBN = ISBNarray.join("\n");
```

Comment:
Cleaning up should be unnecessary as of next version, but if we're doing it, let's join using space, not newline (though it's probably getting converted anyway)